src/number_theory/arithmetic_function.lean

```diff
@@ -221,15 +232,294 @@ instance : semiring (arithmetic_function α) :=
+  intros m n cop,
+  rw sum_mul_sum,
+  symmetry,
+  apply sum_bij (λ (x : (ℕ × ℕ) × ℕ × ℕ) h, (x.1.1 * x.2.1, x.1.2 * x.2.2)),
+  { rintros ⟨⟨a1, a2⟩, ⟨b1, b2⟩⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    rcases h with ⟨⟨rfl, ha⟩, ⟨rfl, hb⟩⟩,
+    simp only [mem_divisors_antidiagonal, nat.mul_eq_zero, ne.def],
+    split, {ring},
+    rw nat.mul_eq_zero at *,
+    apply not_or ha hb },
+  { rintros ⟨⟨a1, a2⟩, ⟨b1, b2⟩⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    rcases h with ⟨⟨rfl, ha⟩, ⟨rfl, hb⟩⟩,
+    dsimp,
+    rw [hf.map_mul_of_coprime cop.coprime_mul_right.coprime_mul_right_right,
+        hg.map_mul_of_coprime cop.coprime_mul_left.coprime_mul_left_right],
+    ring, },
+  { rintros ⟨⟨a1, a2⟩, ⟨b1, b2⟩⟩ ⟨⟨c1, c2⟩, ⟨d1, d2⟩⟩ hab hcd h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at hab,
+    rcases hab with ⟨⟨rfl, ha⟩, ⟨rfl, hb⟩⟩,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at hcd,
+    simp only [prod.mk.inj_iff] at h,
+    ext; dsimp,
+    { transitivity nat.gcd (a1 * a2) (a1 * b1),
+      {rw [nat.gcd_mul_left, cop.coprime_mul_left.coprime_mul_right_right.gcd_eq_one, mul_one]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.1.1, h.1, nat.gcd_mul_left,
+            cop.coprime_mul_left.coprime_mul_right_right.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (a1 * a2) (a2 * b2),
+      {rw [mul_comm, nat.gcd_mul_left, cop.coprime_mul_right.coprime_mul_left_right.gcd_eq_one, mul_one]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.1.1, h.2, mul_comm, nat.gcd_mul_left,
+            cop.coprime_mul_right.coprime_mul_left_right.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (b1 * b2) (a1 * b1),
+      {rw [mul_comm, nat.gcd_mul_right,
+           cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.1, mul_comm c1 d1, nat.gcd_mul_left,
+            cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (b1 * b2) (a2 * b2),
+      {rw [nat.gcd_mul_right,
+           cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.2, nat.gcd_mul_right,
+            cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul] } } },
+  { rintros ⟨b1, b2⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    use ((b1.gcd m, b2.gcd m), (b1.gcd n, b2.gcd n)),
+    simp only [exists_prop, prod.mk.inj_iff, ne.def, mem_product, mem_divisors_antidiagonal],
+    rw [← cop.gcd_mul _, ← cop.gcd_mul _, ← h.1, nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop h.1,
+        nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop.symm _],
+    { rw [nat.mul_eq_zero, decidable.not_or_iff_and_not] at h, simp [h.2.1, h.2.2] },
+    rw [mul_comm n m, h.1] }
+end⟩
```

Comment:
Does it make sense to add `is_multiplicative.pmul` and `.ppow`? Or will those also not be used?

Reply:
I think that's basically how the proof for `is_multiplicative_pow` goes, so I might as well flesh that stuff out.